TMEDIA-459: Shared form component

nschubach opened the pull request:
## Description
Create a reusable form component that we can use for each of the form pages (sign up, log in, etc)

## Jira Ticket
- [TMEDIA-459](https://arcpublishing.atlassian.net/browse/TMEDIA-459)

## Acceptance Criteria
- Reusable form component is validated in Storybook 
- Utilizes form fields 
- Matches the styles of current design documentation (links above)
- Tests created to validate the form functions properly
- Includes: Form title, divider, form inputs, button, divider
- Fits the width of the container

## Test Steps
1. Checkout this branch `git checkout TMEDIA-459-shared-form-component`
2. Run fusion repo with linked blocks `npm run storybook`
3. Validate http://localhost:60001/?path=/story/blocks-identity-components-headlinedsubmitform--basic-form
![image](https://user-images.githubusercontent.com/2287238/131035931-965c8898-f56e-4dac-932c-a4153278629c.png)
4. And http://localhost:60001/?id=blocks-identity-components-headlinedsubmitform--error-form
TODO: insert image of error state

## Effect Of Changes
### Before
Shared form did not exist

### After
It now exists

## Dependencies or Side Effects
Depends on https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067 Button component.  Could potentially be merged with that when the Button is reviewed.

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [x] Confirmed this PR has unit test files
  - [x] Ran `npm run test`, made sure all tests are passing
  - [x] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
feat: Add basic button sizes and primary font for account

Commit:
feat: Add white background button style

Commit:
docs: Add comments

Commit:
docs: Add comment for filter and brightness

nschubach committed:
TMEDIA-459: Shared form component

nschubach committed:
Merge branch 'canary' into TMEDIA-459-shared-form-component

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/index.jsx:
```diff
@@ -0,0 +1,73 @@
+          {buttonLabel}
+        </PrimaryFont>
+        {error ? (
+          <section>
```
Will require `role` to ensure errors are spoken to assistive technologies when added after submission

nschubach replied:
Sorry, yeah, the error state here is a placeholder.  I was waiting on direction as to WHERE it was going to live before actually implementing it properly.  You can ignore this.

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/styles.scss:
```diff
@@ -0,0 +1,40 @@
+.xpmedia-form-filled-button {
+  margin: map-get($spacers, 'md') 0;
+  // primary color for arc subs
+  background-color: #183bb6;
```
Subscriptions designs are using the site's primary color - this color should be getting set from the site property

nschubach replied:
Going to be using Jack's button component for this, so this is also a stand-in to be ignored.

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/styles.scss:
```diff
@@ -0,0 +1,40 @@
+  margin-bottom: map-get($spacers, 'sm');
+
+  .xpmedia-form-field {
+    margin: map-get($spacers, 'md') 0;
```
This will be affected by CSS margin collapse, the bottom margin and the top margin of the next element are the same so you only get the space of one value, not as this reads - the margin between element would be 1.5rem in this case not 3rem as you may expect by setting bottom of one element and top of another

nschubach replied:
I want the collapse and it's intended.

Commit:
fix: Disallow empty text field for button

Commit:
style: Rename class based on feedback

via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r697208545

Commit:
style: Remove stylelint issue with semicolons

Commit:
style: Allow custom attributes based on feedback

https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r697208789

Commit:
docs: Add button type documentation for the button including icon only

Commit:
fix: Use aria label for button

Commit:
feat: Add custom type for button attribute

Commit:
feat: Icon and center align label and text

Commit:
feat: Refactor to simplify props and use proptypes declarations

nschubach committed:
Merge branch 'TMEDIA-402-logged-out-nav-cta' into TMEDIA-459-shared-form-component

nschubach committed:
Actually commit changed code.

nschubach committed:
Fix headline alignment

JackHowa reviewed (approved):
good work!

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/index.jsx:
```diff
@@ -0,0 +1,79 @@
+        {formErrorText ? (
+          <section className="xpmedia-form-error">
+            <PrimaryFont as="p">
+              <svg fill="none" xmlns="http://www.w3.org/2000/svg">
```
This should be in the Engine Theme SDK with all the other icons.

nschubach replied:
Added a spike for this: https://arcpublishing.atlassian.net/browse/TMEDIA-469

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/index.jsx:
```diff
@@ -0,0 +1,79 @@
+          text={buttonLabel}
+        />
+        {formErrorText ? (
+          <section className="xpmedia-form-error">
```
Error's require the attribute `role="alert"` to ensure assistive technologies are aware of an error occurring

nschubach replied:
Will make a PR

nschubach replied:
https://github.com/WPMedia/fusion-news-theme-blocks/pull/1069

matthewroach commented on blocks/identity-block/components/HeadlinedSubmitForm/index.jsx:
```diff
@@ -0,0 +1,79 @@
+        ref={formRef}
+      >
+        {children}
+        <Button
```
Button type needs to be `submit`

nschubach replied:
submit is the default when the button is in a form.

matthewroach replied:
That is only true if the value of the type is not present or invalid - the value being set is `type=button`

nschubach replied:
Added to #1069
